spec/pg_eventstore/subscriptions/basic_runner_spec.rb

```diff
@@ -0,0 +1,689 @@
+        expect { subject; sleep async_shutdown_time + test_adjustment_time }.to change {
+          instance.state
+        }.from("running").to("stopped")
+      end
+      it "releases runner's thread after async_shutdown_time seconds" do
+        expect { subject; sleep async_shutdown_time + test_adjustment_time }.to change {
+          instance.instance_variable_get(:@runner)
+        }.from(instance_of(Thread)).to(nil)
+      end
+      it 'stops runners from processing further' do
+        subject
+        thread = instance.instance_variable_get(:@runner)
+        aggregate_failures do
+          expect(Thread.list).to include(thread)
+          sleep async_shutdown_time
+          expect(Thread.list).not_to include(thread)
+        end
+      end
+    end
+
+    context 'when state is "dead"' do
+      before do
+        callbacks_definitions
+        instance.define_callback(:process_async, :before, proc { raise "You shall not pass!" })
+        instance.start
+        # The thread which spawns to stop the current runner is to fast. It uses #loop method internally - slow it down
+        # a bit to give tests the time to perform assertions
+        allow(instance).to receive(:loop).and_wrap_original do |orig_method, *args, **kwargs, &blk|
+          sleep 0.2
+          orig_method.call(*args, **kwargs, &blk)
+        end
+        instance.wait_for_finish # let the runner die
+      end
+
+      it 'spawns another thread to stop the current runner' do
+        expect { subject }.to change { Thread.list.size }.by(1)
+      end
+      it 'executes :after_runner_stopped action asynchronous' do
+        subject
+        aggregate_failures do
+          expect(after_stopped_task).not_to have_received(:run)
+          sleep async_shutdown_time
+          expect(after_stopped_task).to have_received(:run)
+        end
+      end
+      it 'changes the state to "halting"' do
+        expect { subject }.to change { instance.state }.from("dead").to("halting")
+      end
+      it 'changes the state to "stopped" after async_shutdown_time seconds' do
+        expect { subject; sleep async_shutdown_time }.to change { instance.state }.from("dead").to("stopped")
+      end
+      it "releases runner's thread after async_shutdown_time seconds" do
+        expect { subject; sleep async_shutdown_time }.to change {
+          instance.instance_variable_get(:@runner)
+        }.from(instance_of(Thread)).to(nil)
+      end
+      it 'stops runners from processing further' do
+        subject
+        thread = instance.instance_variable_get(:@runner)
+        sleep 0.1
+        expect(Thread.list).not_to include(thread)
+      end
+    end
+  end
+
+  describe '#restore' do
+    subject { instance.restore }
+
+    let(:before_cbx_task) { double('Before runner restored') }
+    let(:perform_async_task) { double('Async action') }
+    let(:after_error_task) { double('After error happened') }
+    let(:callbacks_definitions) do
```

Comment:
those lets before block for callback could be extracte to the shared context and resued accros specs

Reply:
They could, but it is an overkill - those are just pure doubles. Besides, not every test needs all three